src/appleseed/renderer/modeling/bsdf/microfacetnormalmappinghelper.h

```diff
@@ -0,0 +1,550 @@
+			* G1(perturbed_shading_normal, incoming, original_shading_normal, true);
+
+		pdf_itpo *= lambda_t(perturbed_shading_normal, outgoing, original_shading_normal)
+			* G1(perturbed_shading_normal, incoming, original_shading_normal, true);
+
+		final_value += value_itpo;
+		final_pdf += pdf_itpo;
+	}
+	value = final_value;
+
+	return final_pdf;
+}
+
+template <typename BSDFImpl>
+float MicrofacetNormalMappingHelper<BSDFImpl>::evaluate_pdf(
+        const void*                 data,
+        const bool                  adjoint,
+        const LocalGeometry&        local_geometry,
+        const foundation::Vector3f& outgoing,
+        const foundation::Vector3f& incoming,
+        const int                   modes) const
+{
+	float final_pdf = 0.0f;
+
+	// World space original shading normal.
+	foundation::Vector3f original_shading_normal(local_geometry.m_shading_point->get_original_shading_normal());
+
+	if (foundation::dot(incoming, original_shading_normal) <= 0.0f || foundation::dot(outgoing, original_shading_normal) <= 0.0f)
+		return 0.0f;
+
+	// World space perturbed shading normal.
+	foundation::Vector3f perturbed_shading_normal(local_geometry.m_shading_point->get_shading_normal());
+
+	// Perturbed normal too similar to original. Tangent vector can not be constructed.
+	if (foundation::dot(original_shading_normal, perturbed_shading_normal) > 1.0f - 1e-5)
+		return BSDFImpl::evaluate_pdf(
+			data,
+			adjoint,
+			local_geometry,
+			outgoing,
+			incoming,
+			modes);
+
+	// World space tangent.
+	foundation::Vector3f tangent = build_tangent(original_shading_normal, perturbed_shading_normal);
+
+	// Case: i -> p -> o.
+	if (foundation::dot(incoming, perturbed_shading_normal) >= 0.0f) // cull check for non-adjoint, BSDF::reflective
+	{
+		float pdf_ipo =
+			BSDFImpl::evaluate_pdf(
+			data,
+			adjoint,
+			local_geometry,
+			outgoing,
+			incoming,
+			modes);
+
+		final_pdf += lambda_p(perturbed_shading_normal, outgoing, original_shading_normal)
+			* pdf_ipo
+			* G1(perturbed_shading_normal, incoming, original_shading_normal, true);
+	}
+
+	// Case: i -> p -> t -> o.
+
+	// World space outgoing direction reflected at tangent facet.
+	foundation::Vector3f incoming_reflected =
+		foundation::normalize(incoming - 2.0f * foundation::dot(incoming, tangent) * tangent);
+
+	if (foundation::dot(incoming_reflected, perturbed_shading_normal) >= 0.0f) // cull check for non-adjoint, BSDF::reflective
+	{
+		float pdf_ipto =
+			BSDFImpl::evaluate_pdf(
+				data,
+				adjoint,
+				local_geometry,
+				outgoing,
+				incoming_reflected,
+				modes);
+
+		final_pdf += lambda_p(perturbed_shading_normal, outgoing, original_shading_normal)
+			* pdf_ipto
+			* G1(perturbed_shading_normal, incoming, original_shading_normal, false)
+			* (1.0f - G1(perturbed_shading_normal, incoming_reflected, original_shading_normal, true));
```

Comment:
Small thing, but it bothers me a bit. The order here is different than in line 411. Maybe use the order of the paper, to have a common ground.

Reply:
I agree.
I am using paper order: labda() * brdf() * cosine_mult() * (1-G1) * G1().
I have checked the same in the other code parts but here was the only difference. It is fixed now.